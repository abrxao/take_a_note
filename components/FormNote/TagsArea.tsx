"use client";
import { MouseEventHandler, ReactNode } from "react";
import { Badge } from "../ui/badge";
import { twMerge } from "tailwind-merge";

interface TagsAreaProps {
  tags: string[];
  className?: string;
  clickAction?: MouseEventHandler<HTMLButtonElement>;
  Icon?: any;
}

export default function TagsArea({
  tags,
  className,
  clickAction,
  Icon,
}: TagsAreaProps): ReactNode {
  return (
    <ul className={twMerge("flex flex-wrap gap-2", className)}>
      {tags.map((elem, index) => {
        return (
          <li key={elem + index}>
            <button
              type="button"
              value={elem}
              onClick={clickAction}
              disabled={clickAction == undefined}
              className="hover:opacity-85"
            >
              <Badge className="flex gap-1 items-center pointer-events-none">
                {elem}
                {Icon}
              </Badge>
            </button>
          </li>
        );
      })}
    </ul>
  );
}
