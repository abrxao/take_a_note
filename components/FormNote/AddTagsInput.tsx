"use client";
import { ChangeEventHandler, MouseEventHandler, ReactNode } from "react";
import { Label } from "../ui/label";
import { Input } from "../ui/input";
import { Button } from "../ui/button";
import { Plus } from "lucide-react";
import TagsArea from "./TagsArea";

interface AddTagsInput {
  tags: string[];
  addButtonAction?: MouseEventHandler<HTMLButtonElement>;
  inputOnChange?: ChangeEventHandler<HTMLInputElement>;
  inputValue?: string | number | readonly string[];
  clickOnTagAction?: MouseEventHandler<HTMLButtonElement>;
}

export default function AddTagsInput({
  tags,
  addButtonAction,
  inputOnChange,
  inputValue,
  clickOnTagAction,
}: AddTagsInput): ReactNode {
  return (
    <div className="grid gap-2">
      <Label htmlFor="addTags">Add tags</Label>

      <div>
        <p className="text-xs text-zinc-600 mt-1 dark:text-zinc-300">
          Your tags
        </p>
        <TagsArea
          tags={tags}
          Icon={<Plus size={14} />}
          clickAction={clickOnTagAction}
        />
      </div>
      <div className="relative">
        <Input
          id="addTags"
          placeholder="New Tag"
          onChange={inputOnChange}
          type="text"
          value={inputValue}
          className="my-0"
        />
        <Button
          type="button"
          className="absolute right-0 top-0"
          onClick={addButtonAction}
        >
          <Plus size={24} />
        </Button>
      </div>
    </div>
  );
}
