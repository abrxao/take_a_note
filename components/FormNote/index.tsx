"use client";
import { cn } from "@/lib/utils";
import { Button } from "../ui/button";
import { Input } from "../ui/input";
import { Label } from "../ui/label";
import { z } from "zod";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { Textarea } from "../ui/textarea";
import { useState } from "react";
import AddTagsInput from "./AddTagsInput";
import useUserDataState from "@/stores/userDataStore";
import { AlertTriangle, X } from "lucide-react";
import NoteProps from "@/types/note";
import useNoteDrawerState from "@/stores/noteModalStore";
import { toast } from "sonner";
import Tag from "../Tag";

const createNoteSchema = z.object({
  title: z.string().min(1, "Input a title for you note"),
  description: z.string(),
});

type CreateLoginData = z.infer<typeof createNoteSchema>;

export default function ProfileForm({
  className,
}: React.ComponentProps<"form">) {
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<CreateLoginData>({
    resolver: zodResolver(createNoteSchema),
  });
  const { userTags, addNoteOnStore, editNoteOnStore } = useUserDataState();
  const { editNote, clearEditNote, setIsDrawerNoteOpen } = useNoteDrawerState();
  const [tags, setTags] = useState<string[]>(editNote?.tags ?? []);
  const [currentTag, setCurrentTag] = useState({
    value: "",
    error: "",
  });

  async function addUserNote(data: CreateLoginData) {
    const { title, description } = data;

    try {
      var postJson: NoteProps = {
        title,
        description,
        tags,
        createDate: new Date(),
        favorite: false,
      };
      if (editNote != null) {
        postJson = {
          ...postJson,
          favorite: editNote.favorite,
          id: editNote.id,
          lastEditDate: new Date(),
        };
        editNoteOnStore(postJson);
      } else {
        addNoteOnStore(postJson);
      }

      reset();
      setTags([]);
    } catch (e) {
    } finally {
      if (editNote != null) {
        toast("Note edited");
        clearEditNote();
      } else {
        toast("Note created");
      }
      setIsDrawerNoteOpen(false);
    }
  }

  return (
    <form
      onSubmit={handleSubmit(addUserNote)}
      className={cn("grid items-start gap-4", className)}
    >
      <div className="grid gap-2">
        <Label htmlFor="title">Title *</Label>
        <Input
          aria-label="title input - required"
          defaultValue={editNote?.title ?? ""}
          type="text"
          id="title"
          data-testid="title-input-note"
          register={register("title")}
        />
        {errors.title?.message != undefined && (
          <p className="text-sm dark:text-red-400 flex gap-1 items-center text-red-600">
            {errors.title.message} <AlertTriangle size={14} />
          </p>
        )}
      </div>
      <div className="grid gap-2">
        <Label htmlFor="description">Description *</Label>
        <Textarea
          id="description"
          defaultValue={editNote?.description ?? ""}
          data-testid="description-input-note"
          register={register("description")}
        />
        {errors.description?.message}
      </div>
      <AddTagsInput
        tags={userTags.filter((userTag) => !tags.includes(userTag))}
        inputValue={currentTag.value}
        inputOnChange={(e) =>
          setCurrentTag({ value: e.target.value, error: "" })
        }
        clickOnTagAction={(tag) => {
          var aux = tags;
          /* @ts-ignore */
          aux.push(tag);
          setTags(aux);
          setCurrentTag({
            value: "",
            error: "",
          });
        }}
        addButtonAction={(e) => {
          if (currentTag.value) {
            var aux = tags;
            aux.push(currentTag.value);
            setTags(aux);
            setCurrentTag({
              value: "",
              error: "",
            });
          } else {
            setCurrentTag({
              value: "",
              error: "Coloque um valor valido",
            });
          }
        }}
      />
      {
        <ul className={"flex flex-wrap items-start space-x-2"}>
          {tags.map((elem) => {
            return (
              <li key={elem}>
                <Tag
                  type="button"
                  value={elem}
                  aria-label="remove tag"
                  clickAction={() => {
                    // @ts-ignore
                    const aux = tags.filter((tag) => tag != elem);
                    setTags(aux);
                  }}
                >
                  {elem}
                  <X size={14} />
                </Tag>
              </li>
            );
          })}
        </ul>
      }
      <Button>Save changes</Button>
    </form>
  );
}
