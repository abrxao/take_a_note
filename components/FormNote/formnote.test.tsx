import { act, fireEvent, render, screen } from "@testing-library/react";
import FormNote from ".";

test("simulation input", async () => {
  const { container } = render(<FormNote />);
  const inputTitle = screen.getByTestId("title-input-note");
  const inputDescription = screen.getByTestId("description-input-note");
  const button = screen.getByText("Save changes");
  const titleText = "hello world";
  const descriptionText =
    "Proident anim ad proident anim quis aute laboris amet cupidatat incididunt do occaecat sint.";

  await act(async () => {
    fireEvent.change(inputTitle, { target: { value: titleText } });
    fireEvent.change(inputDescription, { target: { value: descriptionText } });
    fireEvent.click(button);
  });
  expect(inputTitle).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
