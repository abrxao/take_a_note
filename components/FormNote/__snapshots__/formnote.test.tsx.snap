// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simulation input 1`] = `
<div>
  <form
    class="grid items-start gap-4"
  >
    <div
      class="grid gap-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for="title"
      >
        Title *
      </label>
      <input
        aria-label="title input - required"
        class="flex h-10 w-full rounded-md border border-slate-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:bg-slate-950 dark:ring-offset-slate-950 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
        data-testid="title-input-note"
        id="title"
        name="title"
        value=""
      />
    </div>
    <div
      class="grid gap-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for="description"
      >
        Description *
      </label>
      <textarea
        class="flex min-h-[80px] w-full rounded-md border border-slate-200 bg-white px-3 py-2 text-sm ring-offset-white placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:bg-slate-950 dark:ring-offset-slate-950 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300"
        data-testid="description-input-note"
        id="description"
        name="description"
      />
    </div>
    <div
      class="grid gap-2"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for="addTags"
      >
        Add tags
      </label>
      <div>
        <p
          class="text-xs text-zinc-600 mt-1 dark:text-zinc-300"
        >
          Your tags
        </p>
        <ul
          class="flex flex-wrap items-start space-x-2"
        />
      </div>
      <div
        class="relative"
      >
        <input
          class="flex h-10 w-full rounded-md border border-slate-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:bg-slate-950 dark:ring-offset-slate-950 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300 my-0"
          data-testid="input-add-tags"
          id="addTags"
          placeholder="New Tag"
          value=""
        />
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-900 text-slate-50 hover:bg-slate-900/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-10 px-4 py-2 absolute right-0 top-0"
          data-testid="input-add-tags-button"
          type="button"
        >
          <svg
            class="lucide lucide-plus "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
    </div>
    <ul
      class="flex flex-wrap items-start space-x-2"
    />
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-900 text-slate-50 hover:bg-slate-900/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-10 px-4 py-2"
    >
      Save changes
    </button>
  </form>
</div>
`;
